tempesta_fw/ss_skb.c

```diff
@@ -907,13 +918,15 @@ ss_skb_process(struct sk_buff *skb, unsigned int *off, ss_skb_actor_t actor,
 	 */
 	for (i = 0; i < si->nr_frags; ++i) {
 		const skb_frag_t *frag = &si->frags[i];
-		unsigned int frag_size = skb_frag_size(frag);
+		size_t frag_size = skb_frag_size(frag);
 		if (offset < frag_size) {
+			size_t to_read = min(frag_size - offset, max_read);
 			unsigned char *frag_addr = skb_frag_address(frag);
-			*off += frag_size - offset;
-			r = actor(objdata, frag_addr + offset,
-					   frag_size - offset);
-			if (r != SS_POSTPONE)
+
+			*off += to_read;
+			max_read -= to_read;
+			r = actor(objdata, frag_addr + offset, to_read);
+			if ((r != SS_POSTPONE) || !max_read)
 				return r;
 			offset = 0;
 		} else {
```

Comment:
As in comment for `tfw_http_parse_skb()`, it seems the changes are wrong - we should handle the buffering at lower layer.